### PR TITLE
Detect invalid tag lengths in psa_aead_setup

### LINKED ISSUE #5285: PSA AEAD: detect invalid tag lengths during setup
`psa_aead_{encrypt,decrypt}_setup` don't always complain if the tag length encoded in the algorithm is invalid. The goal of this issue is to ensure that they do and that this is tested.

Acceptance: for each of {ChachaPoly, CCM, GCM},  
for each of a reasonable selection of invalid tag lengths (must include at least 0, min_tag_length-1 and, for CCM and GCM, a value between min_tag_length and max_tag_length),  
for each of {encrypt,decrypt}_setup,  
there shall be a passing test case validating that setup fails.

### PR DESCRIPTION
**Pull request by AndrzejKurek**
Read tag lengths from the driver and validate against preset values in `psa_aead_setup`.
Fixes #5285.

### CODE REVIEW

File: tests/suites/test_suite_psa_crypto.data [hunk 1]
```diff
@@ -3611,9 +3643,57 @@ PSA AEAD setup: invalid algorithm (ChaCha20 - Poly1305 with short tag)
...
+
+PSA AEAD setup: AES - CCM, invalid tag length 15
+depends_on:PSA_WANT_ALG_CCM:PSA_WANT_KEY_TYPE_AES
+aead_multipart_setup:PSA_KEY_TYPE_AES:"4189351B5CAEA375A0299E81C621BF43":PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_CCM,15):PSA_ERROR_INVALID_ARGUMENT
+
+PSA AEAD setup: AES - CCM, invalid tag length 17
+depends_on:PSA_WANT_ALG_CCM:PSA_WANT_KEY_TYPE_AES
+aead_multipart_setup:PSA_KEY_TYPE_AES:"4189351B5CAEA375A0299E81C621BF43":PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_CCM,17):PSA_ERROR_INVALID_ARGUMENT
+
+PSA AEAD setup: AES - GCM, invalid tag length 0
```
Comment (mprse) [inline tests/suites/test_suite_psa_crypto.data:3666]: Correct values for GCM: `4, 8, 12, 13, 14, 15, 16`.
We have here the following sections for negative test cases and tested values:
```
1. < 4      : 0, 2, 3
2. <5, 7>   : 
3. <9, 11>  : 11
4. > 16     : 17
```

I think we should test at least the boundary values of each section. So my suggestion is to test the following values:
```
0, 3
5, 7
9, 11
17
```

I think the same approach should be for `CCM` and `ChaCha20-Poly1305`.

**Review by mprse — COMMENTED** [review]
Looks good. Found some minor issues and left 2 comment (code style, testing).

File: tests/suites/test_suite_psa_crypto.data (hunk 1 above)
Comment (AndrzejKurek) [reply tests/suites/test_suite_psa_crypto.data:3666]: That's one of the possible approaches, but you can also see that GCM accepts even numbers apart from `2` and `10`, that's why I added a test for `2`. Now adding for 10 too. As for CCM... if you want to cover every transition boundary, then we'd need 0, 3, 5, 7, 9, 11, 13, 15, 17... That's a bit too much coverage in my opinion, as we can just check the basic principle (even vs odd). As for ChaCha - it accepts only 16, so 0, 15, 17 is fine.

File: library/psa_crypto.c [hunk 2]
```diff
@@ -3819,6 +3819,47 @@ psa_status_t psa_aead_decrypt( mbedtls_svc_key_id_t key,
...
+    {
+        return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+    switch( PSA_ALG_AEAD_WITH_SHORTENED_TAG( alg, 0 ) )
+    {
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)
+        case PSA_ALG_AEAD_WITH_SHORTENED_TAG( PSA_ALG_CCM, 0 ):
+            /* CCM allows the following tag lengths: 4, 6, 8, 10, 12, 14, 16.*/
+            if( tag_len < 4 || tag_len > 16 || tag_len%2 )
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/psa_crypto.c]: Fixed, thanks!

File: tests/suites/test_suite_psa_crypto.data (hunk 1 above)
Comment (mprse) [reply tests/suites/test_suite_psa_crypto.data:3666]: Yes I know that this is a lot of test cases, but usually checking boundary values is the minimum required. 

> ChaCha - it accepts only 16, so 0, 15, 17 is fine.
Yes, if only 16 is valid, then to cover boundary values for negative cases we need `0, 15, 17`.

Maybe someone else should give input here.

Comment (minosgalanakis) [reply tests/suites/test_suite_psa_crypto.data:3666]: @mprse required by what? If that is a formal strong requirement we should do it as defined, regarldess of the overhead. If that is an implied requirement, then we could skip some. I am not very familiar with how much time this adds to testing, but if its minimal I would go for better coverage.

If we are looking to optimize the testing for CCM the valid inputs are `4, 6, 8, 10, 12, 14, 16` and the check condition is 

`if( tag_len < 4 || tag_len > 16 || tag_len%2 )`, it would make sense to check:

0 --> ( true || false || true )
2 --> ( true || false || true )
3 --> ( true || false || false )
5 --> ( false || false|| false ) . This is the equivalent for any of the `5, 7, 9, 11, 13, 15`
17 --> ( false || true|| false )
18 --> ( false || true|| true)

Anything from the accepted numbers covers the (false, false, true) case

I have no strong prefference on the testing choice, so overall the patch looks good to me.

Comment (mprse) [reply tests/suites/test_suite_psa_crypto.data:3666]: I also don't have strong preferences on testing and  I don't know what are the testing requirements in this project. 
From my experience and maybe this is not a best option here (safety-critical software) we were testing at least the boundary values.

Comment (AndrzejKurek) [reply tests/suites/test_suite_psa_crypto.data:3666]: There doesn't seem to be much computation before we exit each negative test case - I'll add more CCM cases.

Comment (AndrzejKurek) [reply tests/suites/test_suite_psa_crypto.data:3686]: Done :)

**Review by mprse — APPROVED** [review]
LGTM

**Review by minosgalanakis — APPROVED** [review]
